crypto/evp/keymgmt_lib.c

```diff
@@ -119,21 +93,75 @@ void *evp_keymgmt_export_to_provider(EVP_PKEY *pk, EVP_KEYMGMT *keymgmt,
                  * forgets to check the return value.
 
                  */
-                if (exportfn(pk->pkeys[j].provdata, &try_import, &import_data)
+                if (exportfn(pk->pkeys[i].provdata, &try_import, &import_data)
                     && (provdata = import_data.provdata) != NULL)
                     break;
             }
         }
+
+        /*
+         * TODO(3.0) Right now, we assume we have ample space.  We will
+         * have to think about a cache aging scheme, though, if |i| indexes
+         * outside the array.
+         *
+         * We assign the ossl_assert() result to i because it has an attribute
+         * saying the result can't go unchecked.
+         *
+         */
+        i = ossl_assert(i < OSSL_NELEM(pk->pkeys));
+    }
+
+    return provdata;
+}
+
+void *evp_keymgmt_export_to_provider(EVP_PKEY *pk, EVP_KEYMGMT *keymgmt,
+                                     int want_domainparams)
+{
+    void *provdata = NULL;
+    size_t i;
+    int domainparams = -1;
+
+    /*
+     * If there is an underlying legacy key and it has changed, invalidate
+     * the cache of provider keys.
+     */
+    if (pk->pkey.ptr != NULL) {
+        /*
+         * If there is no dirty counter, this key can't be used with
+         * providers.
+         */
+        if (pk->ameth->dirty_cnt == NULL)
+            return NULL;
+
+        if (pk->ameth->dirty_cnt(pk) != pk->dirty_cnt_copy)
+            evp_keymgmt_clear_pkey_cache(pk);
     }
 
     /*
-     * TODO(3.0) Right now, we assume we have ample space.  We will
-     * have to think about a cache aging scheme, though, if |i| indexes
-     * outside the array.
+     * See if we have exported to this provider already.
+     * If we have, return immediately.
      */
-    j = ossl_assert(i < OSSL_NELEM(pk->pkeys));
+    for (i = 0;
+         i < OSSL_NELEM(pk->pkeys) && pk->pkeys[i].keymgmt != NULL;
+         i++) {
+        if (domainparams == -1)
+            domainparams = pk->pkeys[i].domainparams;
```

Comment:
Do all entries in pk->pkeys[i] have the same domainparams setting?

Reply:
When I designed this cache, I did intend for it to be able to hold a mixture, but that turned out difficult / flawed. That's why you see these indicators on a per cache item basis, and you're right, it can be taken out from there. I think that's a topic for a separate PR, though...